File: spring-passport/src/main/java/com/oceanheart/passport/repository/SessionRepository.java
package com.oceanheart.passport.repository;

import com.oceanheart.passport.domain.Session;
import com.oceanheart.passport.domain.User;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Repository
public interface SessionRepository extends JpaRepository<Session, UUID> {

    /**
     * Find all sessions for a specific user
     */
    List<Session> findByUser(User user);

    /**
     * Find all sessions for a user by user ID
     */
    List<Session> findByUserId(UUID userId);

    /**
     * Find active sessions for a user (created within the last week)
     */
    @Query("SELECT s FROM Session s WHERE s.user = :user AND s.createdAt > :since")
    List<Session> findActiveSessionsByUser(@Param("user") User user, @Param("since") LocalDateTime since);

    /**
     * Find active sessions for a user by user ID
     */
    @Query("SELECT s FROM Session s WHERE s.user.id = :userId AND s.createdAt > :since")
    List<Session> findActiveSessionsByUserId(@Param("userId") UUID userId, @Param("since") LocalDateTime since);

    /**
     * Find session by ID and user (for security verification)
     */
    Optional<Session> findByIdAndUser(UUID sessionId, User user);

    /**
     * Find session by ID and user ID
     */
    Optional<Session> findByIdAndUserId(UUID sessionId, UUID userId);

    /**
     * Find sessions by IP address
     */
    List<Session> findByIpAddress(String ipAddress);

    /**
     * Find sessions created after a specific date
     */
    List<Session> findByCreatedAtAfter(LocalDateTime date);

    /**
     * Find sessions created between two dates
     */
    List<Session> findByCreatedAtBetween(LocalDateTime start, LocalDateTime end);

    /**
     * Find expired sessions (older than 1 week)
     */
    @Query("SELECT s FROM Session s WHERE s.createdAt < :expireTime")
    List<Session> findExpiredSessions(@Param("expireTime") LocalDateTime expireTime);

    /**
     * Count active sessions for a user
     */
    @Query("SELECT COUNT(s) FROM Session s WHERE s.user = :user AND s.createdAt > :since")
    long countActiveSessionsByUser(@Param("user") User user, @Param("since") LocalDateTime since);

    /**
     * Count total sessions for a user
     */
    long countByUser(User user);

    /**
     * Count total sessions by user ID
     */
    long countByUserId(UUID userId);

    /**
     * Delete all sessions for a user
     */
    @Modifying
    @Transactional
    void deleteByUser(User user);

    /**
     * Delete all sessions for a user by user ID
     */
    @Modifying
    @Transactional
    void deleteByUserId(UUID userId);

    /**
     * Delete expired sessions (cleanup operation)
     */
    @Modifying
    @Transactional
    @Query("DELETE FROM Session s WHERE s.createdAt < :expireTime")
    void deleteExpiredSessions(@Param("expireTime") LocalDateTime expireTime);

    /**
     * Delete sessions created before a specific date
     */
    @Modifying
    @Transactional
    void deleteByCreatedAtBefore(LocalDateTime date);

    /**
     * Delete specific sessions by IDs
     */
    @Modifying
    @Transactional
    @Query("DELETE FROM Session s WHERE s.id IN :sessionIds")
    void deleteByIds(@Param("sessionIds") List<UUID> sessionIds);

    /**
     * Find recent sessions for admin dashboard (last 24 hours)
     */
    @Query("SELECT s FROM Session s WHERE s.createdAt > :since ORDER BY s.createdAt DESC")
    List<Session> findRecentSessions(@Param("since") LocalDateTime since);

    /**
     * Find all sessions ordered by creation date (newest first)
     */
    List<Session> findAllByOrderByCreatedAtDesc();

    /**
     * Find sessions with specific user agent pattern
     */
    @Query("SELECT s FROM Session s WHERE s.userAgent LIKE %:pattern%")
    List<Session> findByUserAgentContaining(@Param("pattern") String pattern);

    /**
     * Get session statistics for admin dashboard
     */
    @Query("SELECT COUNT(s) FROM Session s WHERE s.createdAt > :since")
    long countSessionsCreatedSince(@Param("since") LocalDateTime since);

    /**
     * Find latest session for each user
     */
    @Query("SELECT s FROM Session s WHERE s.createdAt = (SELECT MAX(s2.createdAt) FROM Session s2 WHERE s2.user = s.user)")
    List<Session> findLatestSessionPerUser();

    /**
     * Update session's last accessed time
     */
    @Modifying
    @Transactional
    @Query("UPDATE Session s SET s.updatedAt = :now WHERE s.id = :sessionId")
    void updateLastAccessed(@Param("sessionId") UUID sessionId, @Param("now") LocalDateTime now);
}

File: go-passport/internal/repository/session_repository.go
package repository

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"time"

	"github.com/oceanheart/go-passport/internal/config"
	"github.com/oceanheart/go-passport/internal/models"
)

var (
	ErrSessionNotFound = errors.New("session not found")
)

type SessionRepository struct {
	db *config.Database
}

func NewSessionRepository(db *config.Database) *SessionRepository {
	return &SessionRepository{db: db}
}

func (r *SessionRepository) Create(ctx context.Context, session *models.Session) error {
	query := `
		INSERT INTO sessions (user_id, ip_address, user_agent, created_at, updated_at)
		VALUES ($1, $2, $3, $4, $5)
		RETURNING id`
	
	now := time.Now()
	session.CreatedAt = now
	session.UpdatedAt = now
	
	err := r.db.QueryRowContext(
		ctx,
		query,
		session.UserID,
		session.IPAddress,
		session.UserAgent,
		session.CreatedAt,
		session.UpdatedAt,
	).Scan(&session.ID)
	
	if err != nil {
		return fmt.Errorf("failed to create session: %w", err)
	}
	
	return nil
}

func (r *SessionRepository) FindByID(ctx context.Context, id int64) (*models.Session, error) {
	query := `
		SELECT id, user_id, ip_address, user_agent, created_at, updated_at
		FROM sessions
		WHERE id = $1`
	
	session := &models.Session{}
	err := r.db.QueryRowContext(ctx, query, id).Scan(
		&session.ID,
		&session.UserID,
		&session.IPAddress,
		&session.UserAgent,
		&session.CreatedAt,
		&session.UpdatedAt,
	)
	
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, ErrSessionNotFound
		}
		return nil, fmt.Errorf("failed to find session by ID: %w", err)
	}
	
	return session, nil
}

func (r *SessionRepository) FindByUserID(ctx context.Context, userID int64) ([]*models.Session, error) {
	query := `
		SELECT id, user_id, ip_address, user_agent, created_at, updated_at
		FROM sessions
		WHERE user_id = $1
		ORDER BY created_at DESC`
	
	rows, err := r.db.QueryContext(ctx, query, userID)
	if err != nil {
		return nil, fmt.Errorf("failed to find sessions by user ID: %w", err)
	}
	defer rows.Close()
	
	var sessions []*models.Session
	for rows.Next() {
		session := &models.Session{}
		if err := session.Scan(rows); err != nil {
			return nil, fmt.Errorf("failed to scan session: %w", err)
		}
		sessions = append(sessions, session)
	}
	
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("error iterating rows: %w", err)
	}
	
	return sessions, nil
}

func (r *SessionRepository) Delete(ctx context.Context, id int64) error {
	query := `DELETE FROM sessions WHERE id = $1`
	
	result, err := r.db.ExecContext(ctx, query, id)
	if err != nil {
		return fmt.Errorf("failed to delete session: %w", err)
	}
	
	rowsAffected, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("failed to get rows affected: %w", err)
	}
	
	if rowsAffected == 0 {
		return ErrSessionNotFound
	}
	
	return nil
}

func (r *SessionRepository) DeleteByUserID(ctx context.Context, userID int64) error {
	query := `DELETE FROM sessions WHERE user_id = $1`
	
	_, err := r.db.ExecContext(ctx, query, userID)
	if err != nil {
		return fmt.Errorf("failed to delete sessions by user ID: %w", err)
	}
	
	return nil
}

func (r *SessionRepository) DeleteExpired(ctx context.Context, expiryDuration time.Duration) error {
	query := `DELETE FROM sessions WHERE created_at < $1`
	
	expiryTime := time.Now().Add(-expiryDuration)
	_, err := r.db.ExecContext(ctx, query, expiryTime)
	if err != nil {
		return fmt.Errorf("failed to delete expired sessions: %w", err)
	}
	
	return nil
}

func (r *SessionRepository) Update(ctx context.Context, session *models.Session) error {
	query := `
		UPDATE sessions
		SET ip_address = $1, user_agent = $2, updated_at = $3
		WHERE id = $4`
	
	session.UpdatedAt = time.Now()
	
	result, err := r.db.ExecContext(
		ctx,
		query,
		session.IPAddress,
		session.UserAgent,
		session.UpdatedAt,
		session.ID,
	)
	
	if err != nil {
		return fmt.Errorf("failed to update session: %w", err)
	}
	
	rowsAffected, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("failed to get rows affected: %w", err)
	}
	
	if rowsAffected == 0 {
		return ErrSessionNotFound
	}
	
	return nil
}

func (r *SessionRepository) CountByUserID(ctx context.Context, userID int64) (int64, error) {
	query := `SELECT COUNT(*) FROM sessions WHERE user_id = $1`
	
	var count int64
	err := r.db.QueryRowContext(ctx, query, userID).Scan(&count)
	if err != nil {
		return 0, fmt.Errorf("failed to count sessions: %w", err)
	}
	
	return count, nil
}